src/Nervos/Transaction.cpp

```diff
@@ -0,0 +1,362 @@
+    for (auto& cellDep : cellDeps) {
+        cellDep.encode(cellDepsData);
+    }
+    dataArray.emplace_back(cellDepsData);
+
+    // header deps
+    Data headerDepsData;
+    encode32LE(uint32_t(headerDeps.size()), headerDepsData);
+    for (auto& headerDep : headerDeps) {
+        headerDepsData.insert(headerDepsData.end(), headerDep.begin(), headerDep.end());
+    }
+    dataArray.emplace_back(headerDepsData);
+
+    // inputs
+    Data inputsData;
+    encode32LE(uint32_t(inputs.size()), inputsData);
+    for (auto& input : inputs) {
+        input.encode(inputsData);
+    }
+    dataArray.emplace_back(inputsData);
+
+    // outputs
+    Data outputsData1;
+    std::vector<Data> outputsData1Array;
+    for (auto& output : outputs) {
+        Data outputData1;
+        output.encode(outputData1);
+        outputsData1Array.emplace_back(outputData1);
+    }
+    Serialization::encodeDataArray(outputsData1Array, outputsData1);
+    dataArray.emplace_back(outputsData1);
+
+    // outputs data
+    Data outputsData2;
+    std::vector<Data> outputsData2Array;
+    for (auto& outputData : outputsData) {
+        Data outputData2;
+        encode32LE(uint32_t(outputData.size()), outputData2);
+        outputData2.insert(outputData2.end(), outputData.begin(), outputData.end());
+        outputsData2Array.emplace_back(outputData2);
+    }
+    Serialization::encodeDataArray(outputsData2Array, outputsData2);
+    dataArray.emplace_back(outputsData2);
+
+    Serialization::encodeDataArray(dataArray, data);
+
+    return TW::Hash::blake2b(data, 32, Constants::getHashPersonalization());
+}
+
+Common::Proto::SigningError Transaction::plan(const SigningInput& signingInput) {
+    if (signingInput.amount == 0 && !signingInput.useMaxAmount) {
+        return Common::Proto::Error_zero_amount_requested;
+    }
+    if (signingInput.cells.empty()) {
+        return Common::Proto::Error_missing_input_utxos;
+    }
+
+    cellDeps.emplace_back(Constants::getSecp256k1CellDep());
+    outputsData.emplace_back();
+
+    Common::Proto::SigningError result;
+    if (signingInput.sudtAddress.empty()) {
+        outputs.emplace_back(uint64_t(signingInput.useMaxAmount ? 0 : signingInput.amount),
+                             Script(Address(signingInput.toAddress)), Script());
+        result = planNativeTokenTransfer(signingInput.cells, Address(signingInput.changeAddress),
+                                         signingInput.byteFee);
+    } else {
+        cellDeps.emplace_back(Constants::getSUDTCellDep());
+        outputs.emplace_back(
+            minCellCapacityForSUDT, Script(Address(signingInput.toAddress)),
+            Script(Constants::getSUDTCodeHash(), HashType::Type1, signingInput.sudtAddress));
+        result = planSudtTransfer(signingInput.useMaxAmount ? 0 : signingInput.amount,
+                                  signingInput.cells, Address(signingInput.changeAddress),
+                                  signingInput.byteFee);
+    }
```

Comment:
This seems to be very spaghetti / hard to maintains. What if you have to support more tokens in the future?

consider that one function should have one functionality and use internal details functions if your function address more than one functionality.

Reply:
Done